File: runtime/flutter/foundation/diagnosticableTree.ts
import { IDiagnosticable } from "./diagnosticable";
import { DiagnosticLevel } from "./diagnosticLevel";
import { IDiagnosticsNode } from "./diagnosticsNode";
import { DiagnosticsTreeStyle } from "./diagnosticsTreeStyle";
declare const flutter: {
    foundation: {
        diagnosticableTree: (
            this: void,
            diagnosticableTree: DiagnosticableTree
        ) => DiagnosticableTree;
    };
};
const toStringShallowDefaultProps = {
    joiner: ", ",
    minLevel: DiagnosticLevel.debug,
};
const toStringDeepDefaultProps = {
    minLevel: DiagnosticLevel.debug,
    prefixLineOne: "",
};
export interface IDiagnosticableTree {
    toStringShallow: (props: {
        joiner: string;
        minLevel: DiagnosticLevel;
    }) => string;
    toStringDeep: (props: {
        minLevel: DiagnosticLevel;
        prefixLineOne: string;
        prefixOtherLines?: string | undefined;
    }) => string;
    toStringShort: () => string;
    toDiagnosticsNode: (props: {
        name?: string | undefined;
        style?: DiagnosticsTreeStyle | undefined;
    }) => IDiagnosticsNode;
}
export class DiagnosticableTree implements IDiagnosticable {
    public constructor() {
        flutter.foundation.diagnosticableTree(this);
    }
    private readonly _dart_toStringShallow: (props: {
        joiner: string;
        minLevel: DiagnosticLevel;
    }) => string = undefined as any;
    private readonly _dart_toStringDeep: (props: {
        minLevel: DiagnosticLevel;
        prefixLineOne: string;
        prefixOtherLines?: string | undefined;
    }) => string = undefined as any;
    private readonly _dart_toStringShort: () => string = undefined as any;
    private readonly _dart_toDiagnosticsNode: (props: {
        name?: string | undefined;
        style?: DiagnosticsTreeStyle | undefined;
    }) => IDiagnosticsNode = undefined as any;
    public toStringShallow(props: {
        joiner?: string;
        minLevel?: DiagnosticLevel;
    }): string {
        return this._dart_toStringShallow({
            ...toStringShallowDefaultProps,
            ...props,
        });
    }
    public toStringDeep(props: {
        minLevel?: DiagnosticLevel;
        prefixLineOne?: string;
        prefixOtherLines?: string | undefined;
    }): string {
        return this._dart_toStringDeep({
            ...toStringDeepDefaultProps,
            ...props,
        });
    }
    public toStringShort(): string {
        return this._dart_toStringShort();
    }
    public toDiagnosticsNode(props: {
        name?: string | undefined;
        style?: DiagnosticsTreeStyle | undefined;
    }): IDiagnosticsNode {
        return this._dart_toDiagnosticsNode(props);
    }
}

File: lib/cfr/builtins/libs/flutter/foundation/diagnosticableTree.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

import 'package:hydro_sdk/cfr/builtins/boxing/boxers.dart';
import 'package:hydro_sdk/cfr/builtins/boxing/boxes.dart';
import 'package:hydro_sdk/cfr/builtins/boxing/unboxers.dart';
import 'package:hydro_sdk/cfr/vm/closure.dart';
import 'package:hydro_sdk/cfr/vm/context.dart';
import 'package:hydro_sdk/cfr/vm/table.dart';
import 'package:hydro_sdk/hydroState.dart';

class VMManagedDiagnosticableTree extends VMManagedBox<DiagnosticableTree> {
  VMManagedDiagnosticableTree(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['toStringShallow'] = makeLuaDartFunc(func: (List<dynamic> args) {
      return [
        vmObject.toStringShallow(
            joiner: args[1]['joiner'],
            minLevel: maybeUnBoxEnum(
                values: DiagnosticLevel.values, boxedEnum: args[1]['minLevel']))
      ];
    });
    table['toStringDeep'] = makeLuaDartFunc(func: (List<dynamic> args) {
      return [
        vmObject.toStringDeep(
            minLevel: maybeUnBoxEnum(
                values: DiagnosticLevel.values, boxedEnum: args[1]['minLevel']),
            prefixLineOne: args[1]['prefixLineOne'],
            prefixOtherLines: args[1]['prefixOtherLines'])
      ];
    });
    table['toStringShort'] = makeLuaDartFunc(func: (List<dynamic> args) {
      return [vmObject.toStringShort()];
    });
    table['toDiagnosticsNode'] = makeLuaDartFunc(func: (List<dynamic> args) {
      return [
        maybeBoxObject<DiagnosticsNode>(
          object: vmObject.toDiagnosticsNode(
              name: args[1]['name'],
              style: maybeUnBoxEnum(
                  values: DiagnosticsTreeStyle.values,
                  boxedEnum: args[1]['style'])),
          hydroState: hydroState,
          table: HydroTable(),
        )
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final DiagnosticableTree vmObject;
}

class RTManagedDiagnosticableTree extends DiagnosticableTree
    implements Box<DiagnosticableTree> {
  RTManagedDiagnosticableTree({
    required this.table,
    required this.hydroState,
  }) : super() {
    table['vmObject'] = vmObject;
    table['unwrap'] = makeLuaDartFunc(func: (List<dynamic> args) {
      return [unwrap()];
    });
    table['_dart_toStringShallow'] =
        makeLuaDartFunc(func: (List<dynamic> args) {
      return [
        super.toStringShallow(
            joiner: args[1]['joiner'],
            minLevel: maybeUnBoxEnum(
                values: DiagnosticLevel.values, boxedEnum: args[1]['minLevel']))
      ];
    });
    table['_dart_toStringDeep'] = makeLuaDartFunc(func: (List<dynamic> args) {
      return [
        super.toStringDeep(
            minLevel: maybeUnBoxEnum(
                values: DiagnosticLevel.values, boxedEnum: args[1]['minLevel']),
            prefixLineOne: args[1]['prefixLineOne'],
            prefixOtherLines: args[1]['prefixOtherLines'])
      ];
    });
    table['_dart_toStringShort'] = makeLuaDartFunc(func: (List<dynamic> args) {
      return [super.toStringShort()];
    });
    table['_dart_toDiagnosticsNode'] =
        makeLuaDartFunc(func: (List<dynamic> args) {
      return [
        maybeBoxObject<DiagnosticsNode>(
            object: super.toDiagnosticsNode(
                name: args[1]['name'],
                style: maybeUnBoxEnum(
                    values: DiagnosticsTreeStyle.values,
                    boxedEnum: args[1]['style'])),
            hydroState: hydroState,
            table: HydroTable())
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  DiagnosticableTree unwrap() => this;
  DiagnosticableTree get vmObject => this;
  @override
  String toStringShallow(
      {String joiner = ', ',
      DiagnosticLevel minLevel = DiagnosticLevel.debug}) {
    Closure closure = table["toStringShallow"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }

  @override
  String toStringDeep(
      {DiagnosticLevel minLevel = DiagnosticLevel.debug,
      String prefixLineOne = '',
      String? prefixOtherLines}) {
    Closure closure = table["toStringDeep"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }

  @override
  String toStringShort() {
    Closure closure = table["toStringShort"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }

  @override
  DiagnosticsNode toDiagnosticsNode(
      {String? name, DiagnosticsTreeStyle? style}) {
    Closure closure = table["toDiagnosticsNode"];
    return maybeUnBoxAndBuildArgument<DiagnosticsNode, dynamic>(
        closure.dispatch([table], parentState: hydroState)[0],
        parentState: hydroState);
  }
}

void loadDiagnosticableTree(
    {required HydroState hydroState, required HydroTable table}) {
  table['diagnosticableTree'] = makeLuaDartFunc(func: (List<dynamic> args) {
    return [
      RTManagedDiagnosticableTree(table: args[0], hydroState: hydroState)
    ];
  });
  registerBoxer<DiagnosticableTree>(boxer: (
      {required DiagnosticableTree vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedDiagnosticableTree(
      vmObject: vmObject,
      hydroState: hydroState,
      table: table,
    );
  });
}
